### PR TITLE
Adding PTL test for checking default values of newly created vnode

### PR DESCRIPTION
**Pull request by dinesh1306**
#### Bug/feature Description
*  Automate test for verify pbsnodes -a output on user basis

#### Affected Platform(s)
* All

#### Solution Description
* Added automated test for verify pbsnodes -a attributes on user basis.
    * test_pbsuser_as_user
    * test_pbsuser_as_root

#### Testing logs/output
[result.log](https://github.com/PBSPro/pbspro/files/2177578/result.log)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bhroam** [comment]
Codacy is complaining that you are redefining the built-in 'dict'

**Comment by dinesh1306** [comment]
@bhroam , i have fix Codacy issue.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        if 'sharing' in attr_dict:
+            self.assertEqual(attr_dict['sharing'], 'default_shared')
+
+        if user == 'root':
+            if 'pbs_version' in attr_dict:
+                self.assertEqual(attr_dict['pbs_version'], pbs_version)
+            if 'Port' in attr_dict:
+                self.assertEqual(attr_dict['Port'], '15002')
+
+    def test_pbsnodes_user_basis(self):
```
Comment (shilpakodli) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: Test case name is not matching the test description. Please have two different test cases, one for validating pbs_probe and other is to validate pbs_nodes.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        pbs_version = self.server.pbs_version
+
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        if user == 'root':
+            ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
+        else:
+            ret = self.du.run_cmd(self.server.hostname, command)
+
+        attr_list = ret['out']
```
Comment (shilpakodli) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: Please add a comment mentioning what exactly you are doing in the code.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        host_short_name = self.mom.shortname
+        host_name = self.mom.hostname
+        pbs_version = self.server.pbs_version
+
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        if user == 'root':
+            ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
+        else:
+            ret = self.du.run_cmd(self.server.hostname, command)
```
Comment (shilpakodli) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: This is supposed to run as pbsuser. Pass an extra parameter specifying to run as pbsuser.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            self.assertEqual(attr_dict['resources_assigned.naccelerators'],
+                             '0')
+        if 'resources_assigned.ncpus' in attr_dict:
+            self.assertEqual(attr_dict['resources_assigned.ncpus'], '0')
+        if 'resources_assigned.vmem' in attr_dict:
+            self.assertEqual(attr_dict['resources_assigned.vmem'], '0kb')
+        if 'resv_enable' in attr_dict:
+            self.assertEqual(attr_dict['resv_enable'], 'True')
+        if 'sharing' in attr_dict:
+            self.assertEqual(attr_dict['sharing'], 'default_shared')
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: Also please run this on cray simulator or real CRAY. I think we are missing few more attributes here.
Same thing for UV and ICE.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            self.assertEqual(attr_dict['ntype'], 'PBS')
+        if 'state' in attr_dict:
+            self.assertEqual(attr_dict[ATTR_NODE_state], 'free')
+        if 'resources_available.arch' in attr_dict:
+            self.assertEqual(attr_dict['resources_available.arch'], 'linux')
+        if 'resources_available.vnode' in attr_dict:
+            self.assertEqual(attr_dict['resources_available.vnode'],
+                             host_short_name)
+        if 'resources_available.host' in attr_dict:
+            self.assertEqual(attr_dict['resources_available.host'],
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: where is resources_available.ncpus and pcpus, resources_available.mem?
Also I see 'last_state_change_time' set to date. I suggest just verifying it is set to date format is sufficient.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        if 'sharing' in attr_dict:
+            self.assertEqual(attr_dict['sharing'], 'default_shared')
+
+        if user == 'root':
+            if 'pbs_version' in attr_dict:
+                self.assertEqual(attr_dict['pbs_version'], pbs_version)
+            if 'Port' in attr_dict:
+                self.assertEqual(attr_dict['Port'], '15002')
+
+    def test_pbsnodes_user_basis(self):
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: I suggest splitting the test into 2 for better debugging; test_pbsuser_as_user and test_pbsuser_as_root

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        #converting pbsnodes -a output into dictionary for validation
+        attr_list = ret['out']
+        list_len = len(attr_list) - 1
+        attr_list = filter(None, attr_list)
+        for i in range(1, list_len):
+            attr = attr_list[i].split('=')[0].strip()
+            val = attr_list[i].split('=')[1].strip()
+            attr_dict[attr] = val
+
+        if 'Mom' in attr_dict:
```
[the remark this answers is not in the document]
Comment (bhroam) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: I would suggest using self.assertIn().  It asserts a value is in a dict or list.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class Testpbsnodes_attributes(TestFunctional):
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: We use camelcase for testsuite names, not underscore delimited strings

Comment (agrawalravi90) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: Also, I think it'd be better to make the test suite and file name more generic (like pbs_pbsnodes.py), and instead make the test names more specific.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    def Test_pbsnodes(self, user):
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: If this is not a test function, then please remove the prefix 'Test' and rename it to something more appropriate.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        Function to validate pbsnodes attribute on user basis
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: I'm not sure I understand what this function is doing, is it validating that the default attributes that are set on a vnode are set correctly?

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        """
+
+        attr_dict = {}
+        host_short_name = self.mom.shortname
+        host_name = self.mom.hostname
+        pbs_version = self.server.pbs_version
+
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: Please make the test independent of setUp, delete all existing vnodes and create a new one.

File: test/tests/functional/pbs_pbsnodes_attributes.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        if user == 'root':
+            ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
+        else:
+            ret = self.du.run_cmd(self.server.hostname, command, runas=user)
+
+        #converting pbsnodes -a output into dictionary for validation
+        attr_list = ret['out']
+        list_len = len(attr_list) - 1
+        attr_list = filter(None, attr_list)
+        for i in range(1, list_len):
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes_attributes.py]: Suggestion: instead of going over the list, creating a dictionary from it, and then verifying each attribute explicitly, do the following:
1) Create a dictionary of attributes and their default, expected values.
2) Loop over the list ret['out'], look up the attribute name in the dictionary above and match the attribute value with the one in the dictionary.

That's it, this will reduce the size of your function to less than half, and you'll also be able to validate that no additional attributes were set on the vnode.

**Comment by agrawalravi90** [comment]
Besides the code comments, please also change the title of your PR and the commit to summarize the new test that you are adding. "automate pbsnodes -a test" suggests that there was some kind of manual test previously which is now being automated, and that is not valid for PBSPro open source.

File: test/tests/functional/pbs_pbsnodes_attributes.py (hunk 12 above)
Comment (anamikau) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: it  is a good suggestion. this can be done on certain attributes that do not have more than one value. for the ones that has more than one or vary based on the platform will still need to be on case by case basis.

File: test/tests/functional/pbs_pbsnodes_attributes.py (hunk 11 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: in case of Cray , UV and ICE machine , it is difficult. there are multiple nodes, deleting and creating nodes is very costly.

File: test/tests/functional/pbs_pbsnodes_attributes.py (hunk 10 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: validating values of pbsnodes attributes are correct or not.

Comment (agrawalravi90) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: pbsnodes is a command, so I'm guessing you are talking about node attributes, and since you are not setting any explicitly in the test, it seems like it is testing whether the default values of node attributes are set correctly or not. So, I suggest changing the docstring to say that.

File: test/tests/functional/pbs_pbsnodes_attributes.py (hunk 11 above)
Comment (agrawalravi90) [reply test/tests/functional/pbs_pbsnodes_attributes.py]: Ok, but this test is validating the default values that a node's attributes are set to. So, it is important that this test creates a new vnode and validates its attributes. Is it possible to do that? If not, then I suggest adding skip decorators for those platforms.

**Comment by dinesh1306** [comment]
I have incorporate comments, Please take a look.

**Comment by agrawalravi90** [comment]
I'm afraid the new title "automate pbsnodes -a" is even more misleading than the one before, how about "Adding PTL test for checking default values of pbsnodes -a" ?

**Comment by dinesh1306** [comment]
Thanks @agrawalravi90 , i have renames it.

**Comment by bhroam** [comment]
@dinesh1306 Just changing the PR title isn't enough.  You need to modify your commit title.  Since you are going to squash everything into the first commit, you only need to modify that one.  You do this through git rebase -i with the rename command.  Either the rebase or your wait until you squash and do a git commit --amend, but that is error prone since it requires us to remember to do it.

File: test/tests/functional/pbs_pbsnodes.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        attr_list = ret['out']
+        list_len = len(attr_list) - 1
+        attr_list = filter(None, attr_list)
+        for i in range(1, list_len):
+            attr = attr_list[i].split('=')[0].strip()
+            val = attr_list[i].split('=')[1].strip()
+            attr_dict[attr] = val
+
+        for k in out.items():
+            self.assertIn(out[k], attr_dict[k])
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: use assertequal() instead

File: test/tests/functional/pbs_pbsnodes.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+        ex_reg = "(Mon|Tue|Wed|Thus|Fri|Sat|Sun)\s"
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: typo 'Thurs'

File: test/tests/functional/pbs_pbsnodes.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        pcpus_val = str(pcpus_val)
+        ncpus_val = pcpus_val
+        avail_mem = os.sysconf("SC_PAGE_SIZE")*os.sysconf("SC_PHYS_PAGES")/1024
+        avail_mem = str(avail_mem) + 'kb'
+
+        expect_dict['Mom'] = host_name
+        expect_dict['ntype'] = 'PBS'
+        expect_dict[ATTR_NODE_state] = 'free'
+        expect_dict['resources_available.arch'] = 'linux'
+        expect_dict['resources_available.vnode'] = host_short_name
```
[the remark this answers is not in the document]
Comment (dinesh1306) [reply test/tests/functional/pbs_pbsnodes.py]: The scope of this PR not including Cray and HPE.

**Comment by dinesh1306** [comment]
@anamikau , i have incorporate review comments. please take a look.

File: test/tests/functional/pbs_pbsnodes.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+        attr_dict = {}
+        out = self.expect_pbsnodes_val(TEST_USER)
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        ret = self.du.run_cmd(self.server.hostname, command)
+        attr_list = ret['out']
+        list_len = len(attr_list) - 1
+        attr_list = filter(None, attr_list)
+        for i in range(1, list_len):
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: Why are you starting from index 1?

File: test/tests/functional/pbs_pbsnodes.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+            attr = attr_list[i].split('=')[0].strip()
+            val = attr_list[i].split('=')[1].strip()
+            attr_dict[attr] = val
+
+        for k in out:
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: please use a more descriptive name than 'k', how about 'attr' ?

File: test/tests/functional/pbs_pbsnodes.py (hunk 16 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_pbsnodes.py]: in ret['out'] first element is node name. to split attributes taking starting index 1.

**Comment by agrawalravi90** [comment]
@dinesh1306 Bhroam had a comment about this test being a possible duplicate of another one, please take a look at his comment, thanks.

**Comment by dinesh1306** [comment]
@bhroam , SmokeTest.test_validate_pbsnodes_attrs checks only default attributes exist in pbsnodes output It doesn't validate assigned values of attributes and runs as only PBSROOT.
But the new test case validates pbsnodes' default attributes and also their expected values as ROOT_USER and PBSUSER.

**Comment by anamikau** [comment]
@bhroam we need to run pbsnodes as regular user and as manager as there are few additional attributes that gets listed under root/admin/manager.
Also this is the very reason to add a test for this to make sure we are not making any changes accidentally. And this test is a sanity test of the PBS setup to make sure all the attributes and values are expected after a fresh install of PBS. Don't you want to know if any code change is likely to break pbsnodes?

**Comment by bhroam** [comment]
@anamikau I do want to know if someone breaks pbsnodes.  We already have that test.  We don't need 2 more that basically do the same thing.  I don't think you should force someone to update 3 tests when they add a node attribute.  The added benefit of these tests is not that great in my opinion.

**Comment by agrawalravi90** [comment]
"we need to run pbsnodes as regular user and as manager as there are few additional attributes that gets listed under root/admin/manager."

Running it as manager will verify all attributes. Running it as a normal user will verify a subset of it, how is that useful when we have already verified all the attributes as a manager?

**Comment by anamikau** [comment]
Good point @agrawalravi90 . Now I am thinking test needs to be updated further to make sure that as user you do not see manager permission attributes.

@bhroam yes we do not want duplicate tests and I am totally on board with that. Either this or the other test needs to be obsolete.

**Comment by agrawalravi90** [comment]
"Now I am thinking test needs to be updated further to make sure that as user you do not see manager permission attributes."

That might be tricky to do. When you create a new node, only some of the attributes get set. There are plenty others which might have manager permissions, they are just not set. So, writing a test which does an exhaustive check over all node attributes to make sure that a non-manager user doesn't have access to them would involve first setting all of them using manager user, then stating using normal and making sure the stat gives only a restricted number of attributes.

Unless there's a huge benefit of such a test, I personally think it might not be worth writing, because it's going to be a big pain to write and maintain such a test.

**Comment by anamikau** [comment]
@dinesh1306 there was an offline discussion and below is the conclusion.

- Remove the duplicate test from pbs_smoketest.py
- Add smoke tag for only the test that is running as root/manager
- keep both the tests in same file, may be rename the file as pbs_nodes.py
- update the test to validate the default static values for example, 
Mom = <hostname>
     ntype = PBS
resources_available.arch = <system arch>
     resources_available.host = <host short name>
resources_available.vnode = <host short name>
resv_enable = True

For following values just check they are set and to expected list or format. for example,
last_state_change_time = <date format>
     last_used_time = <date format>
resources_available.ncpus = positive integer

**Comment by bhroam** [comment]
@dinesh1306 before removing the test in SmokeTest, I would review it.  I remember @nithinj having to jump through some hoops to make it work right.  You might also want to work with @npadole20.  She is currently writing a similar test for server attributes.  It would be nice if both tests worked similarly.  This way when someone has to update one of them, they don't have to learn two different ways to do so.

**Comment by anamikau** [comment]
Also please resolve conflicts and travis/appveyor errors.

**Comment by anamikau** [comment]
BTW I did not see anything for sharing attributes as I suggested last time.
Also please resolve conflicts in your branch and travis/appveyor errors.

**Comment by agrawalravi90** [comment]
Marking this as inactive since there has been no activity on this for a month. @dinesh1306 are you still working on this?

**Comment by dinesh1306** [comment]
Hi @agrawalravi90 , i am working on this PR. Please remove the "inactive" tag.

**Comment by dinesh1306** [comment]
@agrawalravi90 , Please review the PR.

File: test/tests/functional/pbs_pbsnodes.py [hunk 18]
```diff
@@ -54,6 +54,70 @@ def setUp(self):
         self.svrname = self.server.pbs_server_name
         self.hostA = self.moms.values()[0].shortname
 
+    def common_setUp(self):
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: please add docstring for this explaining why have you added this

File: test/tests/functional/pbs_pbsnodes.py [hunk 19]
```diff
@@ -107,3 +171,70 @@ def test_pbsnodes_av(self):
...
         self.assertEquals(prev.strip(), now.strip(),
                           'Last used time mismatch after server restart')
+
+    @skipOnCpuSet
+    @skipOnCray
+    def test_pbsnodes_as_user(self):
+        """
+        Validate default values of node attributes for non-root user
+        """
+        self.common_setUp()
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: is this not covered by skipOnCray decorator?

File: test/tests/functional/pbs_pbsnodes.py [hunk 20]
```diff
@@ -107,3 +171,70 @@ def test_pbsnodes_av(self):
...
+    @skipOnCray
+    def test_pbsnodes_as_root(self):
+        """
+        Validate default values of node attributes for root user
+        """
+        self.common_setUp()
+        self.server.manager(MGR_CMD_DELETE, NODE, id="",
+                            sudo=True, expect=True)
+        self.server.manager(MGR_CMD_CREATE, NODE, id=self.mom.shortname)
+        self.server.expect(NODE, {'state': 'free'})
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: just a thought, since your tests needs this special setUp may be above 3 lines can get defined under common_setUp.

File: test/tests/functional/pbs_pbsnodes.py [hunk 21]
```diff
@@ -107,3 +171,70 @@ def test_pbsnodes_av(self):
...
+        # comparing the pbsnodes -a output with expected result
+        for attr in expected_attrs:
+            self.assertEqual(expected_attrs[attr], attr_dict[attr])
+
+        self.verify_node_dynamic_val(attr_dict['last_state_change_time'],
+                                     attr_dict['resources_available.ncpus'],
+                                     attr_dict['pcpus'], attr_dict['sharing'],
+                                     attr_dict['resources_available.mem'])
+
+    @tags('smoke')
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: tag the whole suite with "commands"

**Comment by dinesh1306** [comment]
@anamikau , Ihave incorporate review comments. Please take a look.

File: test/tests/functional/pbs_pbsnodes.py [hunk 22]
```diff
@@ -55,10 +56,17 @@ def setUp(self):
         self.hostA = self.moms.values()[0].shortname
 
     def common_setUp(self):
+        """
+        Common setUp for tests test_pbsnodes_as_user and test_pbsnodes_as_root
+        """
         TestFunctional.setUp(self)
         platform = self.du.get_platform()
         if platform == 'cray' or platform == 'craysim':
             self.skipTest("cannot run on cray or cray_sim system")
```
Comment (anamikau) [inline test/tests/functional/pbs_pbsnodes.py]: I have same question, why not using skiponcray decorator for this?

**Comment by anamikau** [comment]
Also provide the latest execution logs

File: test/tests/functional/pbs_pbsnodes.py (hunk 22 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_pbsnodes.py]: I changed the code to use decorator skipOnCray.

**Comment by dinesh1306** [comment]
@anamikau , please find execution result.
[result.txt](https://github.com/PBSPro/pbspro/files/2740226/result.txt)

File: test/tests/functional/pbs_pbsnodes.py [hunk 23]
```diff
@@ -54,6 +55,74 @@ def setUp(self):
...
+                        'ignore_excl', 'force_excl', 'force_exclhost']
+
+        # verify 'last_state_change_time' matched with datetime
+        last_state_change_time = str(last_state_change_time)
+        if time.strptime(last_state_change_time, "%a %b %d %H:%M:%S %Y"):
+            self.logger.info("last_state_change_time matched with datetime")
+
+        # checking resources_avalable.ncpus and pcpus value are positive value
+        ncpus_val = int(available_ncpus)
+        if ncpus_val >= 0:
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: this if needs an else case to fail the test if the value is not negative. Same goes for pcpus as well.

File: test/tests/functional/pbs_pbsnodes.py [hunk 24]
```diff
@@ -54,6 +55,74 @@ def setUp(self):
...
+            mssg = "resources_available.ncpus have positive int value"
+            self.logger.info(mssg)
+        pcpus_val = int(pcpus)
+        if pcpus_val >= 0:
+            self.logger.info("pcpus have positive int value")
+
+        # verify pcpus and ncpus value are same
+        if pcpus_val == ncpus_val:
+            self.logger.info("pcpus and ncpus have same value")
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: fail the test in the else case here

File: test/tests/functional/pbs_pbsnodes.py [hunk 25]
```diff
@@ -54,6 +55,74 @@ def setUp(self):
...
+
+        # verify node sharing attribute have one of the value in
+        # sharing_val list
+        self.assertIn(sharing, sharing_list)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: Please add an error log message here.

File: test/tests/functional/pbs_pbsnodes.py [hunk 26]
```diff
@@ -54,6 +55,74 @@ def setUp(self):
...
+
+        # checking resources_avalable.mem value is positive value
+        index = available_mem.find('kb')
+        if int(available_mem[:index]) >= 0:
+            mssg = "resources_available.mem have positive int value"
+            self.logger.info(mssg)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: Again, fail the test in the else case here

File: test/tests/functional/pbs_pbsnodes.py [hunk 27]
```diff
@@ -107,3 +176,60 @@ def test_pbsnodes_av(self):
...
+    def test_pbsnodes_as_user(self):
+        """
+        Validate default values of node attributes for non-root user
+        """
+        self.common_setUp()
+        attr_dict = {}
+        expected_attrs = self.get_newnode_attrs(TEST_USER)
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        ret = self.du.run_cmd(self.server.hostname, command)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: Check for the failure of pbsnodes command by verifying that ret['rc'] is 0

File: test/tests/functional/pbs_pbsnodes.py [hunk 28]
```diff
@@ -107,3 +176,60 @@ def test_pbsnodes_av(self):
...
+    def test_pbsnodes_as_root(self):
+        """
+        Validate default values of node attributes for root user
+        """
+        self.common_setUp()
+        attr_dict = {}
+        expected_attrs = self.get_newnode_attrs('root')
+        command = os.path.join(self.server.pbs_conf['PBS_EXEC'],
+                               'bin', 'pbsnodes -a')
+        ret = self.du.run_cmd(self.server.hostname, command, sudo=True)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_pbsnodes.py]: same comment here, check for the failure of pbsnodes command

**Comment by agrawalravi90** [comment]
@dinesh1306 it seems like you are working on this PR again, in which case please remove the WIP tag, I hadn't looked at this PR because of the WIP tag, we use it to tell reviewers "don't review this PR yet"

**Comment by dinesh1306** [comment]
@agrawalravi90 , I have incorporate review commts. Please review.

**Comment by dinesh1306** [comment]
@shilpakodli , @agrawalravi90 , @anamikau Thank you for the Review.
Branch is squashed and rebased, Ready for merge. 
[final_result.txt](https://github.com/PBSPro/pbspro/files/2749609/final_result.txt)

**Comment by anamikau** [comment]
@dinesh1306 something got in before you. please rebase again. Also ask someone in your TZ to do the merge since lots of pending PR's are getting merged now which were blocked due to travis issue.

**Comment by dinesh1306** [comment]
Branch is rebased. Ready fo merge.

**Comment by anamikau** [comment]
Thanks @dinesh1306. merged the changes in.
